### ISSUE TITLE
missing semicolon in doctest has bad recovery (?), parser tries to parse whole module

### ISSUE THREAD

**Issue by jendrikw** [rust-lang/rust#158232]
(please suggest a better wording for the title)

### Code

```Rust
/// ```
/// let mut x = Vec::new();
/// x.push_mut("foo")
/// x.push_mut("bar");
/// ```
pub fn f() {}
```

### Current output

```Shell
---- src/lib.rs - f (line 1) stdout ----
error: expected item, found keyword `let`
 --> src/lib.rs:2:1
  |
2 | let mut x = Vec::new();
  | ^^^ `let` cannot be used for global variables
  |
  = help: consider using `static` and a `Mutex` instead of `let mut`
  = note: for a full list of items that can appear in modules, see <https://doc.rust-lang.org/reference/items.html>
```

### Desired output

```Shell
error: expected `;`, found `x`
 --> src/lib.rs:3:18
  |
3 | x.push_mut("foo")
  |                  ^ help: add `;` here
4 | x.push_mut("bar");
  | - unexpected token
```

### Rationale and extra context

It seems strange that a simple missing semicolon send the parser on a wild path. It seems it wants to parse the code block as a whole module without clear reason, but then only outputs one error message unrelated to the semicolon.

If the last line is removed, the output is (as expected) a type error:

### Other cases

```Rust
---- src/lib.rs - f (line 1) stdout ----
error[E0308]: mismatched types
 --> src/lib.rs:3:1
  |
3 | x.push_mut("foo")
  | ^^^^^^^^^^^^^^^^^ expected `()`, found `&mut &str`
  |
help: consider using a semicolon here
  |
3 | x.push_mut("foo");
  |                  +
help: try adding a return type
  |
1 | fn main() { #[allow(non_snake_case)] fn _doctest_main_src_lib_rs_1_0() -> &mut &str {
  |                                                                        ++++++++++++
```

### Rust Version

```Shell
rustc 1.96.0 (ac68faa20 2026-05-25)
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-05-25
host: x86_64-unknown-linux-gnu
release: 1.96.0
LLVM version: 22.1.2
```

### Anything else?

the suggestion to add a return type to a doctest main could also be improved

**Comment by bjorn3** [comment]
`echo '...' | RUSTDOC_LOG=trace rustdoc --test -` shows:

```
  INFO rustc_interface::passes 0 parse sess buffered_lints
 DEBUG rustdoc::doctest creating test <anon> - f (line 1): let mut x = Vec::new();
 x.push_mut("foo")
 x.push_mut("bar");

running 1 test
 DEBUG rustdoc::doctest::make invalid AST:
 let mut x = Vec::new();
 x.push_mut("foo")
 x.push_mut("bar");
 DEBUG rustdoc::doctest compiler invocation for doctest: UNSTABLE_RUSTDOC_TEST_LINE="1" UNSTABLE_RUSTDOC_TEST_PATH="doctest.rs" "/home/bjorn/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc" "@/tmp/rustdoctestRs1OBb/rustdoc-cfgs" "--edition" "2015" "--target" "x86_64-unknown-linux-gnu" "--color" "never" "--crate-type=bin" "-o" "/tmp/rustdoctestapIzVb/rust_out" "-"
test <anon> - f (line 1) ... FAILED

failures:

---- <anon> - f (line 1) stdout ----
error: expected item, found keyword `let`
 --> doctest.rs:2:1
  |
2 | let mut x = Vec::new();
  | ^^^ `let` cannot be used for global variables
  |
  = help: consider using `static` and a `Mutex` instead of `let mut`
  = note: for a full list of items that can appear in modules, see <https://doc.rust-lang.org/reference/items.html>

error: aborting due to 1 previous error

Couldn't compile the test.

failures:
    <anon> - f (line 1)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

I think what happens is that rustdoc failed to parse the test code as function body and then simply skipped wrapping the doctest in a function, thus asking rustc to literally parse

```rust
let mut x = Vec::new();
x.push_mut("foo")
x.push_mut("bar");
```

as module, for which this is a totally reasonable error message, rather than


```rust
#[test]
fn foo() {
let mut x = Vec::new();
x.push_mut("foo")
x.push_mut("bar");
}
```

as would normally be the case if there was no syntax error.

**Comment by fmease** [comment]
Dupe of https://github.com/rust-lang/rust/issues/142446, it contains a lot more context in the comments, too.
